### PR TITLE
Collect and fit data for acceleration and velocity forces

### PR DESCRIPTION
**Pull request by pkubanek**
- Fit M1M3 velocities and accelerations
- WIP - plotting, interpolating, ..
- fix timestamp usage
- Produce table with both azimuth and elevations
- Fitting values
- was_raised - check if the mirror was raised durin HP measrumenets
- save new inertial compensation tables
- Move functions to AccelerationAndVelocityFitter (criopy)
- Fit azimuth
- Use Pandas Timestamps as indices
- Use full azimth and elevation extends
- Simplyfy residuals storage
- Use progress_apply
- Use progress_apply for calculating mirror forces
- m1m3-aav script
- fix null values
- More functions
- More functions, improved docs
- Documenting the code
- New criopy naming
- fit_values, out_dir command line arguments
- -d command line option
- --load command line argument

### CODE REVIEW

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+from lsst.ts.criopy.m1m3 import AccelerationAndVelocityFitter, ForceCalculator
+from lsst.ts.idl.enums.MTM1M3 import DetailedState
+from lsst_efd_client import EfdClient
+from tqdm import tqdm
+
+tqdm.pandas()
+
+
+class AccelerationAndVelocity:
+    """Compute fir of M1M3 acceleration and velocities.
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: fit

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+    load_from : pathlib.Path
+        Optional path to load data (from hd5 file). If None, data are
+        collected/computed.
+
+    Attributes
+    ----------
+    azimuths : pd.DataFrame
+        Mount azimuth. Contains actual and demand Position, Velocity and
+        Acceleration.
+    calculated : pd.DataFrame
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: maybe calculated_forces?

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+        self.fitter: AccelerationAndVelocityFitter | None = None
+        self.interpolated: pd.DataFrame | None = None
+        self.intervals: pd.DataFrame | None = None
+        self.mirror: pd.DataFrame | None = None
+        self.raw: pd.DataFrame | None = None
+        self.residuals: pd.DataFrame | None = None
+
+        if load_from is not None:
+
+            def try_load(name: str) -> pd.DataFrame | None:
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:109]: is this the right place to put this function? Just wondering

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+                "intervals",
+                "raw",
+                "mirror",
+                "residuals",
+            ]:
+                setattr(self, loads, try_load(loads))
+
+    async def load_detailed_state(self, start_time: Time, end_time: Time) -> None:
+        """Find detailedState. Last pre-dating start_time, and all in
+        start_time till end_time. Those are primary used to verify mirror
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: I am a little confused by what you mean with this comment. Could you clarify it? Once I understand it, maybe I can help with a description that is a little more clear?

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 5]
```diff
@@ -0,0 +1,761 @@
...
+        ----------
+        start : Time
+            Start time. Azimuths and elevations data will be available from
+            that time.
+        end : Time
+            End time. Azimuths and elevations data will be available till that
+            time.
+        """
+
+        async def find_axis(axis: str) -> pd.DataFrame:
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:222]: describe what it does in a comment, please

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 6]
```diff
@@ -0,0 +1,761 @@
...
+            end = index - row["timediff"]
+            if start is not None and (end - start) > pd.Timedelta(seconds=0.5):
+                if self.was_raised(start, end):
+                    ret.append(
+                        [
+                            start,
+                            end,
+                            True,
+                        ]
+                        #    if df_notmoving is None
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: is this supposed to be here?

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+                    applied.my,
+                    applied.mz,
+                    applied.forceMagnitude,
+                ],
+                index=[f"calculated_f{a}" for a in "xyz"]
+                + [f"calculated_m{a}" for a in "xyz"]
+                + ["forceMagnitude"],
+            )
+
+        aav = pd.concat([self.fitter.velocities, self.fitter.accelerations], axis=1)
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: what does aav mean? Maybe you could use a more clear name for the variable?

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 8]
```diff
@@ -0,0 +1,761 @@
...
+
+        logging.info("Calculating new mirror forces - backpropagation")
+        self.calculated = aav.progress_apply(vect_to_fam, axis=1)
+        self.calculated.set_index(self.fitter.aav.index, inplace=True)
+
+    async def load_hardpoints(self, start: Time, end: Time) -> None | pd.DataFrame:
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:364]: add some comment here, please

**Review by gmegh — CHANGES_REQUESTED** [review]
Adding a first round of comments. I will look at it again after some of the clarifications. Thanks!

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 2 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: It's both forces and moments. I made it calculated_fam

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 3 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:109]: Why not? Python allows that, I don't need that function outside of this simple call (yet)

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 5 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:222]: Documented.

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 4 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: Docstring improved.

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 6 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: No. Removed

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 9]
```diff
@@ -0,0 +1,855 @@
...
+
+__all__ = ["AccelerationAndVelocity"]
+
+import argparse
+import asyncio
+import logging
+import os
+import pathlib
+
+import numpy as np
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:30]: Small comment, but should these imports be organized alphabetically?

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+            return ret
+
+        self.elevations = await find_axis("elevation")
+        self.azimuths = await find_axis("azimuth")
+
+    async def find_applicable_times(self) -> None:
+        """Find start and end times of any axis movents. Uses self.elevations
+        and self.azimuths data retrieved in find_az_el method.
+
+        Fills self.intervals with start and stop times usefull for fitting.
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: useful

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 11]
```diff
@@ -0,0 +1,855 @@
...
+                ],
+                self.azimuths[
+                    (abs(self.azimuths.actualVelocity) > self.actual_velocity_limit)
+                    | (abs(self.azimuths.demandVelocity) > self.demand_velocity_limit)
+                ],
+            ]
+        )
+
+        movements.sort_index(inplace=True)
+        movements["timediff"] = movements.index.to_series().diff()
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:300]: Isn't this already available as movements["timediff"]? In line 258 in the previous function you define the timediff, right?

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 12]
```diff
@@ -0,0 +1,855 @@
...
+                        ]
+                    )
+                else:
+                    logging.warning(
+                        f"Interval {start.isot} - {end.isot} "
+                        "mirror was not raised, ignoring."
+                    )
+            start = index
+
+        # if no movement, use full range - this is for short times usefulla ss
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: fix typos

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 13]
```diff
@@ -0,0 +1,855 @@
...
+        # tests
+        if len(ret) == 0:
+            ret = [
+                [
+                    movements.index[0],
+                    movements.index[-1],
+                    True,
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: why do you create this "use" column if it's always True? Can't we just remove it?

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 14]
```diff
@@ -0,0 +1,855 @@
...
+
+        logging.info("Calculating new mirror forces - backpropagation")
+        self.calculated_fam = acceleration_and_velocity.progress_apply(
+            vect_to_fam, axis=1
+        )
+        self.calculated_fam.set_index(self.fitter.aav.index, inplace=True)
+
+    async def load_hardpoints(self, start: Time, end: Time) -> None | pd.DataFrame:
+        """Load hardpoint forces and moments in given interval. Those are used
+        to calculate (throuigh standard distribution matrices) per-actuator
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: through

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py [hunk 15]
```diff
@@ -0,0 +1,855 @@
...
+        """Collect hardpoint forces for interval specified in self.intervals
+        DataFrame.
+
+        Fills self.raw DataFrame.
+        """
+        assert self.azimuths is not None
+        assert self.elevations is not None
+
+        self.raw = pd.DataFrame()
+        for index, row in self.intervals[self.intervals.use].iterrows():
```
Comment (gmegh) [inline python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: see comment on self.intervals.use

**Review by gmegh — COMMENTED** [review]
I think it looks better. Left some minor comments!

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 9 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:30]: There is isort pre-commit hook which takes care of that. But it does order imports as "standard library first". So numpy etc. are on second level. This is correct behaviour.

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 11 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py:300]: But the one from 258 is either for azimuth or elevation. It's correct to compute it again, as this timediff is used to trigger where new slew starts (timediff > 1 sec from previous non-zero movement).

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 13 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: It was originally used to filter out intervals. Dropped now.

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 10 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: And second just few lines under. Fixed.

File: python/lsst/ts/aos/utils/m1m3/m1m3_aav.py (hunk 14 above)
Comment (pkubanek) [reply python/lsst/ts/aos/utils/m1m3/m1m3_aav.py]: Fixed

**Review by gmegh — APPROVED** [review]
Thanks!

**Pull request merged**